File: 23.合并k个升序链表.py
#
# @lc app=leetcode.cn id=23 lang=python
#
# [23] 合并K个升序链表
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        def mergeTwoList(list1, list2):
            dummy = ListNode(0)
            res = dummy
            while list1 is not None and list2 is not None:
                if list1.val < list2.val:
                    dummy.next = list1
                    list1 = list1.next
                else:
                    dummy.next = list2
                    list2 = list2.next
                dummy = dummy.next
            if list1 is not None:
                dummy.next = list1
            elif list2 is not None:
                dummy.next = list2
            return res.next
        
        if len(lists) == 0:
            return None
        
        if len(lists) == 1:
            return lists[0]
        
        i = 0
        tmp = []
        for i in range(0, len(lists) - 1, 2):
            tmp.append(mergeTwoList(lists[i], lists[i + 1]))
        # 处理奇数
        if i + 2 != len(lists):
            tmp.append(lists[-1])
        return self.mergeKLists(tmp)
# @lc code=end

File: 23.合并k个升序链表.cpp
/*
 * @lc app=leetcode.cn id=23 lang=cpp
 *
 * [23] 合并K个升序链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* merge2lists(ListNode* l1, ListNode* l2) {
        if (l1 == nullptr) {
            return l2;
        }
        if (l2 == nullptr) {
            return l1;
        }
        ListNode* tmp = new ListNode(0);
        ListNode* res = tmp;
        while (l1 && l2) {
            if (l1->val < l2->val) {
                tmp->next = l1;
                l1 = l1->next;
            } else {
                tmp->next = l2;
                l2 = l2->next;
            }
            tmp = tmp->next;
        }
        if (l1) {
            tmp->next = l1;
        }
        if (l2) {
            tmp->next = l2;
        }
        return res->next;
    }

    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.size() == 0) {
            return nullptr;
        }
        if (lists.size() == 1) {
            return lists[0];
        }
        if (lists.size() % 2 == 1) {
            lists.push_back(nullptr);
        }
        vector<ListNode*> nt;
        for (int i = 0; i < lists.size(); i+=2) {
            nt.push_back(merge2lists(lists[i], lists[i+1]));
        }
        return mergeKLists(nt);
    }
};
// @lc code=end

